python/activator/activator.py

```diff
@@ -195,6 +227,113 @@ def create_app():
         sys.exit(3)
 
 
+def dict_to_fanned_out_visit(obj, ctx):
+    """
+    Converts object literal(dict) to a Fanned Out instance.
+    Args:
+        ctx (SerializationContext): Metadata pertaining to the serialization
+            operation.
+        obj (dict): Object literal(dict)
+    """
+
+    if obj is None:
+        return None
+
+    return FannedOutVisit(salIndex=obj['salIndex'],
+                          scriptSalIndex=obj['scriptSalIndex'],
+                          groupId=obj['groupId'],
+                          coordinateSystem=obj['coordinateSystem'],
+                          position=obj['position'],
+                          startTime=obj['startTime'],
+                          rotationSystem=obj['rotationSystem'],
+                          cameraAngle=obj['cameraAngle'],
+                          filters=obj['filters'],
+                          dome=obj['dome'],
+                          duration=obj['duration'],
+                          nimages=obj['nimages'],
+                          instrument=obj['instrument'],
+                          survey=obj['survey'],
+                          totalCheckpoints=obj['totalCheckpoints'],
+                          detector=obj['detector'],
+                          private_sndStamp=obj['private_sndStamp'])
+
+
+def keda_start():
+
+    setup_usdf_logger(
+        labels={"instrument": instrument_name},
+    )
+
+    # Initialize local registry
+    registry = activator.repo_tracker.LocalRepoTracker.get()
+    registry.init_tracker()
+
+    # Check initialization and abort early
+    _get_consumer()
+    _get_storage_client()
+    _get_central_butler()
+    _get_local_repo()
+
+    # Initiialize schema registry for fan out
+    fan_out_schema_registry_conf = {'url': fan_out_schema_registry_url}
+    fan_out_schema_registry_client = SchemaRegistryClient(fan_out_schema_registry_conf)
+
+    fan_out_avro_deserializer = AvroDeserializer(schema_registry_client=fan_out_schema_registry_client,
+                                                 from_dict=dict_to_fanned_out_visit)
+
+    fan_out_consumer_conf = {
+        "bootstrap.servers": fan_out_kafka_cluster,
+        "group.id": fan_out_kafka_group_id,
+        "auto.offset.reset": fan_out_kafka_topic_offset,
+        "sasl.mechanism": fan_out_kafka_sasl_mechanism,
+        "security.protocol": fan_out_kafka_security_protocol,
+        "sasl.username": fan_out_kafka_sasl_username,
+        "sasl.password": fan_out_kafka_sasl_password,
+        'enable.auto.commit': False
+    }
+
+    fan_out_consumer = kafka.Consumer(fan_out_consumer_conf, logger=_log)
+
+    _log.info("starting fan out consumer")
+
+    try:
+        fan_out_consumer.subscribe([fan_out_kafka_topic])
+
+        fan_out_listen_start_time = time.time()
+
+        while time.time() - fan_out_listen_start_time < fanned_out_msg_listen_timeout:
+            fan_out_consumer.subscribe([fan_out_kafka_topic])
```

Comment:
Is this code really supposed to re-subscribe every 5 seconds?

Reply:
I'm also a bit confused about the workflow. Something calls `keda_start`, and then... an actual nextVisit might arrive 300 seconds (by default) later? Does that mean this function is being run (and rerun) in the absence of messages?

Reply:
Thanks for catching.  A redundant subscribe.  I will remove and sorry if that lead to the confusion. 

Keda is watching the consumer group lag on the fanned out event Kafka topic.   Keda will create new Kubernetes Jobs/Pods based on the consumer group lag and the scaling parameters (similar to the Knative scaling).   Some of the pods will be created ahead of when the next visit event arrives.  There needs to be a poll for messages to account for that.

There is a maximum run time in Keda, but the maximum run time by itself would have had Pods prematurely terminated.   For example with a 15 minute Keda timeout a Pod could be polling for 10 minutes then receives a fan out message leaving only 5 minutes to complete processing.  Providing a message polling time ensures we can set a maximum run time in Keda.   300 seconds is a guess on what to start with for the timer and will probably be adjusted down.  The Keda maximum run time is set for the message polling time + time to run processing.

Reply:
Just to be sure I understood that explanation: a Kubernetes job is booted and runs `keda_start`, but the job is not actually "assigned" a specific nextVisit until it reads the next one from the topic? In other words, the Keda framework is responsible only for starting/closing jobs, but not actually assigning work?

I understand re: the run time -- it's the same distinction as between Knative timeout and Gunicorn timeout, and we probably want the same solution in the long run (the config specifies the processing timeout, and the framework timeout is extrapolated from that).

Reply:
Correct.